TriangleGenome
==============
README
TriangleGenome
Hans Week
Paige Romero
Ben Packer



philosophy of this program
This program uses lots of seeding and hillclimbing to reach good images. The hillclimbing is threaded to make it faster. 1 of 3 sample
methods (discussed further below) genereates an images with no triangle overlap by arranging them on a grid. We find that this yields lots
of small level detail when crossed over with more organic seeding methods.

Important parameters
The most important parameters for this project are: the number of tribes, the size of each tribe (number of Genomes in each tribe),
the number of steps of generations before inter-tribe Cross over is called and the number of generations before intra-tribe crossover is called.
The number of tribes with best results is 2-8, though the program will accept values up to 1000. The number of Genomes within a tribe for best 
performance is 10.

The logic of the program
When program is executed, the GUI appears and Tribes are created for the default starting image (Mona Lisa). If selecting new images,
the program will recalculate/reseed the tribes after run is pressed. When run is pressed, control goes to each tribe thread.
When the threads are run, the tribe class' climbRoutine() method calls goToLocalMax() N times (N is given by NBREEDSTEPS).
In goToLocalMax a thread is created for each Genome's HillClimb object. When run, this object selects a gene in the genome to mutate, and 
an element or "allele" in that gene to mutate. It then selects a value to mutate it by, checks to make sure this mutation will have valid
values, then calls mutate on it. It checks the fitness of the Genome before and after the mutation with a scaled-image fitness-comparison.
If the mutation is found to be good (fitness value has decreased) then HillClimb calls that mutation in a loop until it makes fitness worse
then it stops mutating that gene (after it removes the final, unfit mutation). The parameters that can change in mutation are:
1 x value of a vertex, 1 y value of a vertex, any of the three color channels r,g,b plus the alpha (opacity) channel. In addition, all three
x values can change at once, as can all three y values (this moves the triangle while preserving its shape) and finally, the "layer" of a gene
can be changed (it trades places with another gene above or below it). After the specified number of generations has occurred in hill climb,
the hill climber object calls the tribes interCross method, which has crosses over genomes within the tribe. The logic of interCross is, it selects
random genomes, but favors the more fit ones (we use a gaussian random for this). It then crosses over the two selected using single point
cross over, but crosses them 6 times. If the number of resulting genomes is larger than the specified maximum (ENDINGTRIBESIZE), then the least fit
genomes are removed. Calls to crossover are made to preserve diversity, and so genomes are tested for Hamming distance, and a crossover is rejected
on close Genomes. After inter-tribe crossover happens, the tribe returns to hillclimbing. Intra-tribe crossover is called as a function of generations
as well, but less often than inter-tribe crossover. When intra-cross is called, it is called in the triangleGenomeGUI object's crossTribes method.
it pauses all the tribe threads, has each tribe provide deep copies of 2 gaussian-random Genomes. Intra cross takes the two Genomes from each
tribe, puts them in a list and crosses them over repeatedly. The resulting Genomes are then sorted, and the ones yielding greatest fitness are
added to the tribes randomly. Then the tribes returns to hillclimb after intra Cross.

Additional features
The program can write the currently selected genome to an XML file as well as display in table format. When saving an file, the
user will be prompted for a filename. The file will be saved to the TriangleGenome folder which contains the src folder***********.
The program can also read these saved XML files, and will build a Genome from it and add it to the 0th index of the currently
selected Tribe.

Seeding conditions:
This program chooses from 3 methods of seeding: sample seeding, average seeding and grid arrangement seeding. Sample seeding selects s color pixel contained
within the bounds of each triangle and assigns that color to the whole triangle. Average seeding averages the entire picture to choose triangle colors and
grid seeding lays out the triangles on a grid so they do not overlap. 

notes of caution
the Pause button may fail to pause come threads during immediately, leaving ghost updates for a few moments. The program also slows down
while calculating crossovers, preventing momentary suspension of statistic updates.

Project breakdown by team mate:
Ben Packer: worked on the GUI and the XML read/write/table display
Paige Romero: worked on crossover, gene/genome structure, genome utilities, Hillclimbing, statistics, GUI, tribe, scaling/displaying images, seeding Genomes, multithreading of Hillclimber
Hans Weeks: worked on Gene/Genome structure, Genome utilities, Hamming distance, hillclimbing and mutation, tribe, UnitTesting, XML read/write, Crossover class and inter/intra crossover, seeding genomes and multithreading tribes
